lib/gc/Analysis/MatmulConfigAnalysis.cpp

```diff
@@ -0,0 +1,422 @@
+                        linalg::LinalgOp &linalgOp, ArrayRef<uint32_t> shape,
+                        SystemDesc &sysDesc, const CostModelFn &costModel,
+                        float eliminationRatio = 0.5, float threshold = -1) {
+  std::vector<MatmulConfig> result;
+  std::vector<float> costs;
+  std::vector<size_t> idx;
+  for (auto [i, config] : llvm::enumerate(configs)) {
+    costs.push_back(costModel(linalgOp, shape, config, sysDesc));
+    idx.push_back(i);
+  }
+  std::stable_sort(idx.begin(), idx.end(), [&costs](size_t i1, size_t i2) {
+    return costs[i1] < costs[i2];
+  });
+  double thresholdCost =
+      costs[idx[(size_t)(eliminationRatio * configs.size())]];
+  thresholdCost =
+      threshold < thresholdCost && threshold > 0 ? threshold : thresholdCost;
+  for (size_t i = 0; i < configs.size(); i++) {
+    if (costs[idx[i]] <= thresholdCost) {
+      result.push_back(configs[idx[i]]);
+    }
+  }
+  LLVM_DEBUG(llvm::dbgs() << "thresholdCost is: " << thresholdCost
+                          << "\nbest with cost: " << costs[idx[0]] << "\n"
+                          << configs[idx[0]] << "\n worst with cost: "
+                          << costs[idx[configs.size() - 1]] << "\n"
+                          << configs[idx[configs.size() - 1]] << "\n");
+  return !result.empty() ? result : configs;
+}
+
+// prepare the config candidates
+std::vector<MatmulConfig>
+prepareConfigCandidates(Operation *root, SystemDesc &sysDesc,
+                        ArrayRef<uint32_t> shape,
+                        ArrayRef<uint32_t> givenInnermostBlock) {
+  std::vector<MatmulConfig> configs;
+  uint32_t threads = sysDesc.getNumThreads();
+  std::vector<uint32_t> MThreadsCandidates =
+      getCandidate((uint32_t)threads, 1U);
+  std::vector<uint32_t> NThreadsCandidates =
+      getCandidate((uint32_t)threads, 1U);
+  std::vector<uint32_t> KThreadsCandidates =
+      getCandidate((uint32_t)threads, 1U);
+  uint32_t noSmallBlockNeedThreshold = 8 * 8U;
+  std::vector<uint32_t> MBlockCandidates = getCandidate(
+      (uint32_t)shape[0], shape[0] > noSmallBlockNeedThreshold ? 8U : 1U,
+      (uint32_t)shape[0]);
+  std::vector<uint32_t> NBlockCandidates =
+      getCandidate((uint32_t)shape[1],
+                   shape[1] > noSmallBlockNeedThreshold ? 8U : 1U, shape[1]);
+  std::vector<uint32_t> KBlockCandidates =
+      getCandidate((uint32_t)shape[2],
+                   shape[2] > noSmallBlockNeedThreshold ? 8U : 1U, shape[2]);
+  std::vector<uint32_t> innerMostMBlockCandidates = getCandidate(
+      (uint32_t)shape[0], shape[0] > noSmallBlockNeedThreshold ? 8U : 1U, 256U);
+  std::vector<uint32_t> innerMostNBlockCandidates = getCandidate(
+      (uint32_t)shape[1], shape[1] > noSmallBlockNeedThreshold ? 8U : 1U, 256U);
+  std::vector<uint32_t> innerMostKBlockCandidates = getCandidate(
+      (uint32_t)shape[2], shape[2] > noSmallBlockNeedThreshold ? 8U : 1U, 256U);
+  if (givenInnermostBlock.size() == 3) {
+    innerMostMBlockCandidates =
```

Comment:
Better to change the logic to only calculate the candidates when `givenInnermostBlock.size() != 3` or its value is not 0?

Reply:
Changed